### DISCUSSION TITLE
Question about mqnic_ptp logic.

### DISCUSSION THREAD

**Discussion by hsincsu** [corundum/corundum#183] [Q&A] [answered]
I am confused about the ptp_td_phc and ptp_td_leaf. I find that the mqnic_ptp_clock set tod timestamp to ptp_td_phc and read tod timestamp from ptp_td_leaf. And ptp_td_phc only share ptp_clk,ptp_sample_clk and ptp_td_sdo wire with ptp_td_leaf. So i test the setting tod timestamp function with cocotb simulation.But it seems the setting tod doesn't influence the ptp_td_leaf 's output.

**Comment by alexforencich** [comment, marked as answer]
It does affect it, but it will take decent number of clock cycles to update.  The serial data is sent in 256 bit frames, and the information to set the ToD clock is sent once every three frames, so could possibly need to wait up to maybe 1000 clock cycles or so to see the ToD output of the leaf clock update.

Don't forget to observe the AXI stream ready-valid handshake; ptp_td_phc will not accept a ToD update on every cycle, only in certain internal states.  You can follow the update process: first your new ToD setting will get loaded into ts_tod_s_reg/ts_tod_ns_reg in ptp_td_phc, and ts_tod_updated_reg will be set.  The rel-ToD ns offset will also be computed.  This will get sent in message 0 based on td_msg_i_reg.  Then, the new ToD value will be deserialized into dst_tod_ns_shadow_reg/dst_tod_s_shadow_reg, and dst_tod_step_shadow_reg should also get set.  This will trigger the new ToD time to be loaded immediately into ts_tod_s_reg/ts_tod_ns_reg, where it will appear at the output.

Note that the exact value might not ever show up at the output; setting the ToD is a non-precision operation with potentially large unpredictable latency, so you would then use a precise relative step or a slew to synchronize the clock.

**Reply by hsincsu** [reply]
Thanks, Alex. This helped me a lot. I think I should wait for more clock cycles to check the output. I'm sure the value has been written into the ptp_td_phc's ts_tod_s_reg/ts_tod_ns_reg. I set the ToD value for some reason. I have equipment that can output two synchronized clocks over a long distance, maybe hundreds of kilometers. I want to use two FPGAs with PTP enabled connected by these two clocks for mqnic_ptp. But I have to reset them at the same time using the PPS signal supplied by that equipment.

**Reply by alexforencich** [reply]
Bringing in an external PPS signal is presumably going to need some sort of a servo loop to maintain synchronization, not a simple reset.  I'm thinking it might be a good idea to make a "timing I/O" module that can be used to facilitate this sort of thing - receive an external timing signal like PPS or IRIG or some such, capture an appropriate timestamp, and then hand that off to a servo loop somewhere.

**Reply by hsincsu** [reply]
Thanks for the suggestion! I think that's my next step to implement. Currently, I just reset them to the same time to count which is not a really synchronization. But resetting them to the same time on the rising edge of PPS may keep them synchronized in some way? Because if one FPGA restarts, it may start at zero, then I will use the reset function to sync the FPGA's time to the other one.
